File: archives/harvest.js
const feedUrl = 'https://www.freeconferencecall.com/rss/podcast?id=2dd4f6a755aa45d0e05e72cc2367b2611992a141827eb6addeed79c5baf445fe_292812442';
const sessionsFilename = 'sessions.json';
const feedFilename = 'rss.xml';
const downloadsDirName = 'downloads';
const bucketPrefix = 'https://oapodcasts.s3.amazonaws.com/';

const fs = require('fs');
const https = require('https');
const promisify = require('util').promisify;
const parseXmlString = require('xml2js').parseString;
const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);
const Podcast = require('podcast');

const safeFilename = str => str.replace(/[:/]/g, '_');

const extractItems = json => {
  const items = json.rss.channel[0].item;
  var newitems = [];
  for (const item of items) {
    const newitem = {
      title: item.title.join(' '),
      // link: item.link.join(' '),
      sourcelink: item.enclosure[0].$.url,
      link: encodeURI(bucketPrefix + safeFilename(item.title.join(' ') + '.mp3')),
      pubDate: new Date(item.pubDate.join(' ')),
      // description: item.description.join(' ')
      description: 'Please visit www.ObjectivismSeminar.com for more information.',
    }
    newitems.push(newitem)
  }

  return newitems;
};

function fetchXmlFeedItems(url, callback) {
  const req = https.request(url, function (res) {
    var xml = '';

    res.on('data', function (chunk) {
      xml += chunk;
    });
    res.on('end', function () {
      parseXmlString(xml, function (err, result) {
        callback(null, extractItems(result));
      });
    });

    res.on('error', function (e) {
      callback(e, null);
    });
    res.on('timeout', function (e) {
      callback(e, null);
    });
  });
  req.end();
}
const fetchFeedItems = promisify(fetchXmlFeedItems)

var callbackDownload = function(url, dest, progress, cb) {
  var file = fs.createWriteStream(dest);
  var request = https.get(url, function(response) {
    response.pipe(file);
    file.on('finish', function() {
      file.close(cb);  // close() is async, call cb after close completes.
    });
    if (progress) { 
      let count = 0;
      response.on('data', chunk => progress(count += chunk.length));
    }
  }).on('error', function(err) { // Handle errors
    fs.unlinkSync(dest); // Delete the file async. (But we don't check the result)
    if (cb) cb(err.message);
  });
};
const download = promisify(callbackDownload);


const buildFeed = sessions => {
  const feed = new Podcast({
    title: "The Objectivism Seminar",
    description: "A weekly online conference call to systematically study the philosophy of Objectivism via the works of prominent Rand scholars.",
    feedUrl: "https://www.objectivismseminar.com/archives/rss",
    siteUrl: "https://www.objectivismseminar.com/",
    imageUrl: "https://www.objectivismseminar.com/assets/images/atlas-square.jpg",
    author: "Greg Perkins, Host",
    itunesOwner: { name: "Greg Perkins", email: "greg@ecosmos.com" },
    itunesExplicit: true,
    itunesCategory: [{text: "Society & Culture", subcats: [{text: "Philosophy", subcats: []}]}],
    language: "en",
    pubDate: new Date(),
  });

  sessions.forEach(session => {
    feed.addItem({
      title: session.title,
      description: session.description,
      
      url: session.link,
      link: session.link,
      category: 'Philosophy',
      guid: session.link,
      date: session.pubDate,
      enclosure: {
        url: session.link,
      }
    });
  });

  return feed;
}


try {
  main()
} catch (e) {
  console.log('MAIN ERROR:', e);
}


async function main() {
  const skipDownloads= process.argv.length > 2 && process.argv[2] == "skip";
  const [feedItems, sessionData] = await Promise.all([fetchFeedItems(feedUrl), readFile(sessionsFilename)]);
  const sessions = JSON.parse(sessionData);
  for (const session of sessions) {
    session.pubDate = new Date(session.pubDate);
  }

  var newItems = []
  for (const item of feedItems) {
    if (!sessions.find(x => x.link === item.link)) {
      newItems.push(item);
    }
  }

  if (!newItems.length) {
    console.log('no new sessions');
  }

  if (!fs.existsSync(downloadsDirName)) {
    fs.mkdirSync(downloadsDirName);
  }

  if (!skipDownloads) {
  for (const item of newItems) {
    await download(item.sourcelink, downloadsDirName + '/' + safeFilename(item.title + '.mp3'), count => {
      process.stdout.cursorTo(0);
      process.stdout.write(`${item.title} ==> ${count} `);
    });
    console.log('complete');
    delete item.sourcelink;
  }
  }

  const updatedSessions = [...newItems, ...sessions];

  const feed = buildFeed(updatedSessions);

  await Promise.all([writeFile(sessionsFilename, JSON.stringify(updatedSessions, null, 2)), writeFile(feedFilename, feed.buildXml())]);
  console.log('wrote sessions.json and rss feed files')
}

File: archives/harvest.py
#!./.venv/bin/python

import errno
import feedparser
import json
import os
from podgen import Podcast, Media, Episode, Category, Person
import re
import requests
import urllib.request
import urllib.parse
import time

default_description = "Please visit www.ObjectivismSeminar.com for more information."
sessions_filename = 'sessions.json'
rss_filename = 'rss.xml'
download_dir = "./downloads"
bucket_prefix = 'https://oapodcasts.s3.amazonaws.com/'
# ipfs_prefix = 'https://cloudflare-ipfs.com/ipfs/'
ipfs_prefix = 'https://gateway.pinata.cloud/ipfs/'
ipfs_suffix = '/audio.mp3'
feedUrl = ('https://www.freeconferencecall.com/rss/podcast' +
           '?id=2dd4f6a755aa45d0e05e72cc2367b2611992a141827eb6addeed79c5baf445fe_292812442')


def safe_name(title):
    return re.sub(r'[:/]', '_', title)


def safe_pinata_path(title):
    return re.sub(r'[:/.]', '_', title)


def copyfileobj(fsrc, fdst, callback, length=16*1024):
    copied = 0
    while True:
        buf = fsrc.read(length)
        if not buf:
            break
        fdst.write(buf)
        copied += len(buf)
        callback(copied)


print('>>> fetching FreeConferenceCall.com rss feed')
feed_items = feedparser.parse(feedUrl)['entries']

print('>>> loading existing sessions.json file')
session_items = json.load(open(sessions_filename))

new_items = []
for item in feed_items:
    link_info = next((x for x in item['links'] if x['type'] == 'audio/mp3'), None)
    if not next((x for x in session_items if item['title'] == x['title']), None):
        new_items.append({
            'title': item['title'],
            'sourcelink': link_info['href'],
            'length': int(link_info['length']),
            'description': default_description,
            'pubDate': time.strftime('%Y-%m-%dT%H:%M:%SZ', item['published_parsed'])
        })

print(f'>>> identified {len(new_items)} new session(s) by title')  # TODO: do this by pubDate?

# ensure downloads directory is in place
if not os.path.exists(download_dir):
    try:
        os.makedirs(download_dir)
    except OSError as exc:  # Guard against race condition
        if exc.errno != errno.EEXIST:
            raise

# download new items
for item in new_items:
    url = item.pop('sourcelink')
    length = item['length']
    title = item['title']

    def progress(bytes_read):
        print(f'>>> {title} -- {int(100 * bytes_read/length)}% downloaded', end='\r')

    file_path = f'{download_dir}/{safe_name(title)}.mp3'

    try:
        if os.path.getsize(file_path) == length:
            print(f'>>> {title} -- 100% downloaded')
            continue
    except FileNotFoundError:
        pass

    print(f'>>> {title} -- 0% downloaded', end='\r')
    with urllib.request.urlopen(url) as response, open(file_path, 'wb') as out_file:
        copyfileobj(response, out_file, progress)
        print(f'>>> {title} -- 100% downloaded')

# updload new items to pinata (ipfs pinning service)
for item in new_items:
    length = item['length']
    title = item['title']
    file_path = f'{download_dir}/{safe_name(title)}.mp3'
    myheaders = {
        "pinata_api_key": os.environ['PINATA_API_KEY'],
        "pinata_secret_api_key": os.environ['PINATA_SECRET_API_KEY']
    }
    myfile = {
        # Using the title as an IPFS directory containing audio.mp3 because
        # stupid iTunes just HAS to see an .mp3 at the end of a URL.
        "file": (f'{safe_pinata_path(title)}{ipfs_suffix}', open(file_path, 'rb'))
    }

    print(f'>>> uploading {title}...', end='\r')
    response = requests.post('https://api.pinata.cloud/pinning/pinFileToIPFS',
                             files=myfile,
                             headers=myheaders)
    print(f'>>> uploaded {title}       ')
    if not response.ok:
        raise Exception(f'upload failed for {title}', response.text)

    cid = response.json()['IpfsHash']
    item['CID'] = cid
    item['GUID'] = cid
    if next((x for x in session_items if x['CID'] == cid), None):
        print(f'WARNING: duplicate CID {cid} for new item: {title}')

# write the new sessions json file
updated_session_items = new_items + session_items

for item in updated_session_items:
    item['link'] = f'{ipfs_prefix}{item["CID"]}{ipfs_suffix}'

with open(sessions_filename, 'w') as outfile:
    json.dump(updated_session_items, outfile, indent=2)

print('>>> wrote fresh sessions.json file')

# write the new rss file
p = Podcast()

p.name = "The Objectivism Seminar"
p.category = Category("Society &amp; Culture", "Philosophy")
p.language = "en-US"
p.explicit = True
p.description = ("A weekly online conference call to systematically study " +
                 "the philosophy of Objectivism via the works of prominent Rand scholars.")
p.website = "https://www.objectivismseminar.com"
p.image = "https://www.objectivismseminar.com/assets/images/atlas-square.jpg"
p.feed_url = "https://www.objectivismseminar.com/archives/rss"
p.authors = [Person("Greg Perkins, Host", "greg@objectivismseminar.com")]
p.owner = Person("Greg Perkins", "greg@ecosmos.com")

p.episodes += [Episode(title=x['title'],
                       media=Media(x['link'], type="audio/mpeg", size=x['length']),
                       id=x['GUID'],
                       publication_date=x['pubDate'],
                       summary=x['description']) for x in updated_session_items]

p.rss_file(rss_filename)

print('>>> wrote fresh rss.xml file')
